Add support for grouped transactions

oliviera9 opened the pull request.

gskapka commented:
Before I review this, as discussed in our call, please provide a helper method for creating these types of transactions.

Also, I notice there are commits in this PR that make the same changes as commits in the other PR. They should be removed/cleaned up. If you need that same code, just rebase this branch onto the other one.

oliviera9 commented:
I rebased from `feat/application-call` (changing the base to that branch would inhibit GH frorm running CI), added the helper function, and a new test in which I group an ASA transfer and an app call, and verify the concatenated bytes (these bytes are obtained from the js sdk as usual).
One thing I did not understand is that I had to add the `Copy` trait to `AlgorandHash`.
Otherwise rustc complains:
```
borrow of partially moved value: `tx2`
partial move occurs because `tx2.group` has type `std::option::Option<algorand_hash::AlgorandHash>`, which does not implement the `Copy` traitrustc[E0382](https://doc.rust-lang.org/error-index.html#E0382)
transactions_group.rs(211, 30): `tx2.group` partially moved due to this method call
transactions_group.rs(211, 30): consider calling `.as_ref()` to borrow the type's contents: `as_ref().`
```.
Are you OK with the new trait? Could you explain me in simple words what the motivation behind this error?

gskapka commented on src/algorand_transactions/transactions_group.rs:
```diff
@@ -0,0 +1,220 @@
+            return Err(AlgorandError::Custom("Empty transactions list".to_string()));
+        }
+        if txns.len() > Self::MAX_TX_GROUP_SIZE {
+            return Err(AlgorandError::Custom("Too much transaction".to_string()));
```
```suggestion
            return Err(AlgorandError::Custom("Too  many transactions in group!".to_string()));
```

gskapka commented on src/algorand_transactions/transactions_group.rs:
```diff
@@ -0,0 +1,220 @@
+            .map(|ref msg_pack_bytes| Self::prefix_group_byte(msg_pack_bytes))
+    }
+
+    pub(crate) fn concat_signed_transactions(txns: &[AlgorandSignedTransaction]) -> Result<String> {
```
This should be `pub` not `pub(crate)`.

Also, this is the final hex for pushing to a node right? The user doesn't need to know or care that it's concatting the txs. So let's name it similarly to the fxn call we make to the underlying txs:

```suggestion
     pub fn to_hex(txns: &[AlgorandSignedTransaction]) -> Result<String> {
```

gskapka replied:
Also, when re-architectured per my main review comment, this fxn signature will become:

```suggestion
    pub fn to_hex(&self) -> Result<String> {
```

oliviera9 replied:
`to_hex` no longer exists after refactoring.

gskapka commented on src/algorand_transactions/transactions_group.rs:
```diff
@@ -0,0 +1,220 @@
+        );
+        let gid = TxGroup::assign_group_id(&mut [&mut tx1, &mut tx2]);
+        let expected_group = [
+            227, 122, 130, 133, 152, 152, 220, 61, 244, 82, 93, 115, 121, 163, 19, 112, 44, 228,
```
Let's use hex for things like this so that it's more concise and more readable.

oliviera9 replied:
Are you happy with this?
```
        assert_eq!(hex::encode(gid.unwrap().to_bytes()), expected_group);
```

gskapka replied:
Works for me!

gskapka commented on src/algorand_transactions/transactions_group.rs:
```diff
@@ -0,0 +1,220 @@
+        prefix
+    }
+
+    pub fn encode_for_signing(&self) -> Result<Bytes> {
```
This name makes no sense. We never sign anything here. The txs included in the group are already signed, and we don't sign over the group ID hash either, so this needs renaming so as not to be confusing.

oliviera9 replied:
I renamed it to `encode_for_hashing`, as the resulting hash will be the group id.

gskapka replied:
Perfect :+1:

gskapka commented on src/algorand_transactions/transactions_group.rs:
```diff
@@ -0,0 +1,220 @@
+        state.serialize_field("txlist", &self.tx_group_hashes)?;
+        state.end()
+    }
+}
```
This manual impl of `Serialize` is doing exactly the same thing as `    #[serde(rename = "txlist", default)]` on L#14, meaning you can just derive `Serialize` and remove this manual impl entirely.

gskapka commented on src/algorand_transactions/transactions_group.rs:
```diff
@@ -0,0 +1,220 @@
+        if txns.len() > Self::MAX_TX_GROUP_SIZE {
+            return Err(AlgorandError::Custom("Too much transaction".to_string()));
+        }
+        let mut ids: Vec<AlgorandHash> = vec![];
+        for t in txns {
+            ids.push(t.to_raw_tx_id()?);
+        }
```
Let's try and be as explicit with naming as possible to make the lives of future readers easier:

```suggestion
        let mut tx_ids: Vec<AlgorandHash> = vec![];
        for tx in txns {
            tx_ids.push(tx.to_raw_tx_id()?);
        }
```

Also, there's no need for a mutable variable and a loop here. Just map over the txs:

```rust
let ids = txns.iter().map(AlgorandTransaction::to_raw_tx_id).collect::<Result<Vec<AlgorandHash>>>()?;
```

oliviera9 replied:
Compiler does not give me to pass `AlgorandTransaction::to_raw_tx_id` to `map`, and I think it is because `to_raw_tx_id` needs to be applied to a struct.
For instance, this works:
```
        let tx_ids = txns.iter().map(|tx| tx.to_raw_tx_id()).collect::<Result<Vec<AlgorandHash>>>()?;
```

gskapka replied:
^ no worries! It's doing the exact same thing, I just thought we might be able to call it first class for less visual noise, but alas!

gskapka commented:
Per our discussion @oliviera9, you pointed out that the group ID changes the tx hash of the transactions, and thus must be assigned before a tx is signed, hence the current, mutable architecture.

I still propose to fix the mess of mutability via much the same way as outlined above, but now we need `TxGroup::new` to take a vec of _unsigned_ txs. Then we can do all the same stuff, and provide a method to sign the tx and get the final output, which method will take a list of private keys of the same length as the number of txs contained. OR, if the list of private keys is of length one, we can assume the user wants all txs signed with the same key.

oliviera9 commented on src/algorand_transactions/transactions_group.rs:
```diff
@@ -0,0 +1,240 @@
+                group_id: None,
+                tx_group_hashes: res,
+            }),
+            _ => Err(AlgorandError::Custom("error".to_string())),
```
Maybe it would be better to propagate the error: `Err(e) => Err(e)`

gskapka replied:
Yeah, this one needs changing for sure. See my review for more details (once I've finished it).

oliviera9 commented on src/algorand_transactions/transactions_group.rs:
```diff
@@ -0,0 +1,240 @@
+
+    pub fn compute_group_id(&self) -> Result<AlgorandHash> {
+        if self.tx_group_hashes.is_empty() {
+            return Err(AlgorandError::Custom("Empty transactions list".to_string()));
```
Another usage of `AlgorandError::Custom`...

gskapka replied:
See my review (once done).

oliviera9 commented on src/algorand_transactions/transactions_group.rs:
```diff
@@ -0,0 +1,240 @@
+                .enumerate()
+                .map(|(i, tx)| tx.sign(keys[i]))
+                .collect(),
+            _ => Err(AlgorandError::Custom("error1".to_string())),
```
Another usage of `AlgorandError::Custom`

gskapka replied:
See my review (once done).

gskapka commented on src/algorand_transactions/transaction.rs:
```diff
@@ -301,10 +301,20 @@ impl AlgorandTransaction {
+    pub fn group(&self) -> Option<AlgorandHash> {
+        self.group
+    }
+
```
Might be better here to make a more useful helper, since `group` is already `pub` in this struct, so a user can just get this exact result via `tx.group` instead of using this fxn `tx.group()`. So it's redundant. 

Instead, let's return a result, so that we can help the user get the underlying group hash:

```suggestion
    pub fn group(&self) -> Result<AlgorandHash> {
        match self.group {
            Some(hash) => Ok(*hash),
            None => Err("No group ID set in transaction!".into()),
        }
    }

```

gskapka commented on src/algorand_transactions/transactions_group.rs:
```diff
@@ -0,0 +1,240 @@
+                group_id: None,
+                tx_group_hashes: res,
+            }),
+            _ => Err(AlgorandError::Custom("error".to_string())),
```
Variables like this `a` which is not named anything useful nor is it used > 1 time anywhere else is a code smell. As is `res` when not in a test. Let's be more descriptive. What is `res`? It's `raw_tx_ids`. So let's call it that instead.

In fact, let's refactor this to remove the pointless `a` variable entirely, and skip the need to match on anything:

```suggestion
        transactions
            .iter()
            .map(|tx| tx.to_raw_tx_id())
            .collect::<Result<Vec<AlgorandHash>>>()
            .map(|raw_tx_ids| Self {
                group_id: None,
                txs: transactions.to_vec(),
                tx_group_hashes: raw_tx_ids,
            })
```

You'll notice this also has the added benefit of us not having to deal with the errors, because any errors that are generated in `to_raw_tx_id` will automatically get bubbled up, since we're dealing with `Result`s here.

__NOTE:__ See a later comment where I point out that we should be assigning the group IDs _in this function_ since we have all the information required to do so.  In fact, that also means you can drop the `Option` wrapper on the group ID in this struct, because you can just initialize with `AlgorandHash::default()`, then replace it with the correctly calculated one. This is great for the end user, because then all they have to do is create a `GroupTx` from their txs, and sign it. Easy peasy.

oliviera9 replied:
Cool! Thanks again for the hints

gskapka commented on src/algorand_transactions/transactions_group.rs:
```diff
@@ -0,0 +1,240 @@
+        }
+    }
+
+    pub fn assign_group_id(&mut self) -> Result<AlgorandHash> {
+        let group_id = TxGroup::compute_group_id(self)?;
+        self.txs = self
+            .txs
+            .iter()
+            .map(|tx| tx.assign_group_id(group_id))
+            .collect();
+        self.group_id = Some(group_id);
+        Ok(group_id)
+    }
```
This should not really even exist. 

We have ALL the information to assign the group IDs correctly when we _ build_ this struct.

gskapka replied:
(See note on the `new()` function. Let's remove this & do everything in there please!)

oliviera9 replied:
Moved inside `new()` as you suggested.
Tests reflect the change as they just create the txs group with `new` and sign the transactions!

gskapka commented on src/algorand_transactions/transactions_group.rs:
```diff
@@ -0,0 +1,240 @@
+            .join(""))
+    }
+
+    pub fn sign_transactions(&self, keys: &[&AlgorandKeys]) -> Result<String> {
```
Same story here as with `new()`. Let's use the `Result`s to handle the errors for us! 

```rust
    pub fn sign_transactions(&self, keys: &[AlgorandKeys]) -> Result<String> {
        let num_keys = keys.len();
        match num_keys {
            1 => {
                // NOTE: We assume that all txs are to be signed with this one key...
                self.txs
                    .iter()
                    .map(|tx| tx.sign(keys[0]))
                    .collect::<Result<Vec<AlgorandSignedTransaction>>>()
            },
            num_keys if num_keys == self.txs.len() => {
                // NOTE: We assume that there's a key for each tx, so we use those...
                self.txs
                    .iter()
                    .zip(keys.iter())
                    .map(|(tx, key)| tx.sign(&key))
                    .collect::<Result<Vec<AlgorandSignedTransaction>>>()
            },
            _ => Err(format!(
                "Please provide either ONE private key, or {} private keys! (IE: one for each tx!)",
                num_keys
            )
            .into()),
        }
        .and_then(Self::to_hex)
    }
```

__NOTE:__ See how we don't need to use the `AlgorandError` enum explicitly, we can just create a string, pass it to the `Err(..)` variant of the `Result` enum, and call `.into()` and let the compiler figure it out. Neat.

__NOTE:__ Also note that I've added some notes for our sake, when we read this again in a few months time and wonder what the heck we were doing :rofl:

__NOTE:__ We don't need the double reference in the parameter either. We can take a reference to an array of keys, not a reference to an array of references to the keys. That's a bit too abstract.

oliviera9 replied:
Invaluable hints! Many thanks

oliviera9 replied:
Also the `zip()` function! At the beginning I was wondering how I could iterate through keys and txs at the same time, and this is simply perfect.

gskapka commented on src/algorand_transactions/transactions_group.rs:
```diff
@@ -0,0 +1,246 @@
+
+    pub fn compute_group_id(&self) -> Result<AlgorandHash> {
+        if self.tx_group_hashes.is_empty() {
+            return Err(AlgorandError::Custom("Empty transactions list".to_string()));
```
Not that it requires changing, but as I mentioned elsewhere, we can do exactly the same thing but more succinctly like so:

```suggestion
            return Err("Empty transactions list".into());
```

The compiler will figure out it's for the `Custom` variant.

gskapka reviewed (changes requested):
LGTM - just needs the minor version number bumping in the `Cargo.toml` please!

gskapka reviewed (approved):
:boom: LGTM!
